### PR TITLE
introduced ParsedZeroAverageConstraints

### PR DESCRIPTION
**Pull request by asartori86**

### CODE REVIEW

**Comment by asartori86** [comment]
@ESeNonFossiIo puoi dare un'occhiata ai test `fixed_lines_00` `01` `02` perché non passano?

**Comment by luca-heltai** [comment]
Why **NullAverage** and not **ZeroAverage**?

**Comment by asartori86** [comment]
> Why NullAverage and not ZeroAverage?

:) non sono bravo con i nomi

**Comment by asartori86** [comment]
cambio?

**Comment by luca-heltai** [comment]
Ho una preferenza per **ZeroAverage**, ma se preferisci l'altro, lascia pure... :)

**Comment by asartori86** [comment]
domani mattina cambio :) ora vado a dormire.. spero

File: include/deal2lkit/parsed_zero_average_constraints.h [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+
+  /**
+   * return the ComponentMask
+   */
+  ComponentMask get_mask () const;
+
+
+  /**
+   * declare_parameters is inherithed by ParameterAcceptor
```
Comment (luca-heltai) [inline include/deal2lkit/parsed_zero_average_constraints.h:114]: This _has_ to be virtual, then...

File: include/deal2lkit/parsed_zero_average_constraints.h [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+   */
+  void declare_parameters (ParameterHandler &prm);
```
Comment (luca-heltai) [inline include/deal2lkit/parsed_zero_average_constraints.h]: same here

File: include/deal2lkit/parsed_zero_average_constraints.h [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+  /**
+   * parse_parameters_call_back is inherithed by ParameterAcceptor
+   */
+  void parse_parameters_call_back ();
```
Comment (luca-heltai) [inline include/deal2lkit/parsed_zero_average_constraints.h]: and here

File: include/deal2lkit/parsed_zero_average_constraints.h [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+  ParsedZeroAverageConstraints  (const std::string &name,
+                                 const unsigned int &n_components=1,
+                                 const std::string &component_names = "",
+                                 const std::string &default_components = "",
+                                 const std::string &default_boundary_components ="");
+
+
+  /**
+   * This function must be called in order to apply the constraint on
+   * the domain.
```
Comment (luca-heltai) [inline include/deal2lkit/parsed_zero_average_constraints.h]: What happens if I call this function and no domain components where specified? I'd expect it to simply leave untouched the constraint matrix. If this is so, why do we need two different functions? Can't we just do everything on one function?

File: include/deal2lkit/parsed_zero_average_constraints.h [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+   */
+  void apply_zero_average_constraints(const DoFHandler<dim,spacedim> &dof_handler,
```
Comment (luca-heltai) [inline include/deal2lkit/parsed_zero_average_constraints.h:97]: General note for documentation: leave out all unneeded sentences... 

"This function must be called" -> "Compute the zero average constraints on the boundary, and apply them on the given constraint matrix".

Same above.

File: include/deal2lkit/parsed_zero_average_constraints.h [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+   *
+   * - a list of components, separated by a "," (i.e. 0,p,5), that are
+   *   to be constrained on the whole domain
+   *
+   * - a list of components, separated by a "," (i.e. 0,p,5), that are
+   *   to be constrained on the boundary
+   */
+  ParsedZeroAverageConstraints  (const std::string &name,
+                                 const unsigned int &n_components=1,
+                                 const std::string &component_names = "",
```
Comment (luca-heltai) [inline include/deal2lkit/parsed_zero_average_constraints.h:88]: The default should be compatible with the number of components: "u" is ok.

Comment (asartori86) [inline include/deal2lkit/parsed_zero_average_constraints.h:88]: preferirei lasciarlo vuoto perché se è vuoto aggiunge in automatico tante 'u' quante sono le componenti

Comment (luca-heltai) [inline include/deal2lkit/parsed_zero_average_constraints.h:88]: (Y)

**Comment by asartori86** [comment]
PR https://github.com/dealii/dealii/pull/2062 is required

**Comment by asartori86** [comment]
currently, travis fails because the deal.II version is not updated

**Comment by asartori86** [comment]
```
/home/travis/build/mathLab/deal2lkit/build/lib/libdeal2lkit.g.so: error: undefined reference to 'void dealii::DoFTools::extract_dofs<2, 3>(dealii::DoFHandler<2, 3> const&, dealii::ComponentMask const&, std::vector<bool, std::allocator<bool> >&)'
```

**Comment by luca-heltai** [comment]
Can you remove the cache, and make it update, or does it download from my dealii build?

**Comment by asartori86** [comment]
serve la PR https://github.com/dealii/dealii/pull/2062

**Comment by asartori86** [comment]
ancora non è stata mergiata

**Comment by luca-heltai** [comment]
si che e' stata mergiata...

**Comment by asartori86** [comment]
any comments?

File: include/deal2lkit/parsed_zero_average_constraints.h [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+                                 const std::string &default_components = "",
+                                 const std::string &default_boundary_components ="");
+
+
+  /**
+   * Compute the zero average constraints on the boundary, and apply
```
Comment (luca-heltai) [inline include/deal2lkit/parsed_zero_average_constraints.h]: This is not only on the boundary now...

**Comment by luca-heltai** [comment]
fix the comment above, and merge.
:+1:

**Pull request merged**